src/Akka.Persistence.MongoDb/Journal/MongoDbJournal.cs

```diff
@@ -290,7 +290,36 @@ protected override async Task<IImmutableList<Exception>> WriteMessagesAsync(IEnu
                 NotifyNewEventAppended();
             return result;
         }
+        private async ValueTask InsertEntries(IEnumerable<JournalEntry> entries)
+        {
+            if (_settings.Transaction)
+            {
+                var sessionOptions = new ClientSessionOptions { };
+                using (var session = await _client.StartSessionAsync(sessionOptions/*, cancellationToken*/))
+                {
+                    // Begin transaction
+                    session.StartTransaction();
+                    try
+                    {
+                        foreach (var entry in entries)
+                        {
+                            await _journalCollection.Value.InsertOneAsync(session, entry);
```

Comment:
`InsertManyAsync` does not support transaction?

Reply:
We should implement that then - need to get rid of 1 by 1

Reply:
Per our call this morning, you need to cite your sources as to why it should be done this way. I won't approve the PR without it.

Reply:
Can you give us a website link we can read?

Reply:
From one of the links you included:

> Starting in version 4.2,
> MongoDB creates as many oplog entries as necessary to the encapsulate all write operations in a transaction, instead of a single entry for all write operations in the transaction. This removes the 16MB total size limit for a transaction imposed by the single oplog entry for all its write operations. Although the total size limit is removed, each oplog entry still must be within the BSON document size limit of 16MB.

Each individual document must be under the 16mb size limit, but that's per-document - not per-transaction. This was different prior to Mongo 4.2.

Mongo 4.2 shipped in 2019 - we should assume that anyone using the MongoDb plugin is running on that version or later going forward (or they're using an older version of this plugin.)

Reply:
@eaba change this back to `InsertManyAsync` with a transaction - that appears to be the winner based on the material you've shared with us. Thank you for doing that - that made the decision making possible.